README.md
[![Build Status](https://travis-ci.org/skaji/Process-Pipeline.svg?branch=master)](https://travis-ci.org/skaji/Process-Pipeline)

# NAME

Process::Pipeline - execute processes as pipeline

# SYNOPSIS

In shell:

    $ zcat access.log.gz | grep -v 127.0.0.1 | grep -c POST

In perl5:

    use Process::Pipeline;

    my $pipeline = Process::Pipeline->new
      ->push(sub { my $p = shift; $p->cmd("zcat", "access.log.gz")   })
      ->push(sub { my $p = shift; $p->cmd("grep", "-v", "127.0.0.1") })
      ->push(sub { my $p = shift; $p->cmd("grep", "-c", "POST"       });

    my $result = $pipeline->start;
    if ($result->is_success) {
       my $fh = $result->fh; # output filehandle of $pipeline
       say <$fh>;
    }

In perl5 with DSL:

    use Process::Pipeline::DSL;

    my $pipeline = proc { "zcat", "access.log.gz"   }
                   proc { "grep", "-v", "127.0.0.1" }
                   proc { "grep", "-c", "POST"      };

    my $result = $pipeline->start;
    if ($result->is_success) {
       my $fh = $result->fh; # output filehandle of $pipeline
       say <$fh>;
    }

# DESCRIPTION

Process::Pipeline helps you write a pipeline of processes.

# MOTIVATION

It is known that we should avoid shell-invocation in perl.
But, because the notation of shell is very convenient,
I sometimes find myself invoking shell. Oops.

The main reason for invoking shell in perl is
that perl does not have as convenient notation as shell has.

Process::Pipeline try to give an easy pipeline notation to perl. Why don't you change

    chomp(my $num = `zcat access.log.gz | grep -v 127.0.0.1 | grep -c POST`);

into

    use Process::Pipeline::DSL;
    my $p = proc { "zcat", "access.log.gz"   }
            proc { "grep", "-v", "127.0.0.1" }
            proc { "grep", "-c", "POST"      };
    my $r = $p->start;
    if ($r->is_success) {
      my $fh = $r->fh;
      chomp(my $num = <$fh>);
    }

# METHODS

## new

    my $pipeline = Process::Pipeline->new;

Constructor.

## push

    $pipeline->push(sub {
      my $p = shift;
      $p->cmd("zcat", "access.log.gz");
    });
    $pipeline->push(sub {
      my $p = shift;
      $p->set("2>", "/dev/null");
      $p->cmd("zcat", "access.log.gz");
    });

Push a Process::Pipeline::Process object to the pipeline.

## start

    my $result = $pipeline->start;

Start the pipeline. It returns a Process::Pipeline::Result object.

    my $result = $pipeline->start;
    my $bool   = $reuslt->is_success; # all commands exit successfully
    my $fh     = $reuslt->fh;         # pipeline's output filehandle

# DSL

There is a DSL for Process::Pipeline. Process::Pipeline::DSL exports
`proc` and `set` functions, and you can construct pipelines easily.

    use Process::Pipeline::DSL;
    my $p = proc { "git", "archive", "--format=tar", "--prefix=repo/", "HEAD" }
            proc { set ">" => "repo.tar.gz"; "gzip" };
    my $r = $p->start;

# COPYRIGHT AND LICENSE

Copyright 2016 Shoichi Kaji <skaji@cpan.org>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Process-Pipeline

{{$NEXT}}

0.04  2016-03-21 12:45:39 JST
    - Close unneccesary filehandles

0.03  2016-03-20 23:48:58 JST
    - Save tempfile until all processes exit (#2 eserte)
    - Fix variable names in pod (#3 eserte)

0.02  2016-03-20 21:59:09 JST
    - State explicitly that this module does not support MSWin32 (#1 chorny)

0.01  2016-03-20 12:26:45 JST
    - Initial release


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.043.
use strict;
use warnings;

use Module::Build 0.28;
use lib qw{inc}; use MyBuilder;

my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "execute processes as pipeline",
  "dist_author" => [
    "Shoichi Kaji <skaji\@cpan.org>"
  ],
  "dist_name" => "Process-Pipeline",
  "dist_version" => "0.03",
  "license" => "perl",
  "module_name" => "Process::Pipeline",
  "recursive_test_files" => 1,
  "requires" => {
    "Process::Status" => 0,
    "perl" => "5.008001"
  },
  "test_requires" => {
    "Test::More" => "0.98"
  }
);


my %fallback_build_requires = (
  "Module::Build" => "0.28",
  "Test::More" => "0.98"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = MyBuilder->new(%module_build_args);


$build->create_build_script;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.043.
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "execute processes as pipeline",
  "AUTHOR" => "Shoichi Kaji <skaji\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Process-Pipeline",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008001",
  "NAME" => "Process::Pipeline",
  "PREREQ_PM" => {
    "Process::Status" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => "0.98"
  },
  "VERSION" => "0.01",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Process::Status" => 0,
  "Test::More" => "0.98"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Process/Pipeline/DSL.pm
package Process::Pipeline::DSL;
use 5.22.1;
use experimental qw/ postderef signatures /;
use Process::Pipeline;

use Exporter qw/ import /;
our @EXPORT = qw/ proc set /;

our $PIPELINE;
our $PROCESS;

sub set ($key, $value = undef) {
    die "Cannot call outside proc()\n" unless $PROCESS;
    $PROCESS->set($key, $value);
}

sub proc ($code, @process) :prototype(&;@) {
    if (!$PIPELINE) {
        local $PIPELINE = Process::Pipeline->new;
        local $PROCESS  = Process::Pipeline::Process->new;
        $PROCESS->cmd($code->());
        $PIPELINE->_push($PROCESS);
        $PIPELINE->_push($_) for map { $_->{process}->@* } @process;
        return $PIPELINE;
    } else {
        local $PROCESS = Process::Pipeline::Process->new;
        $PROCESS->cmd($code->());
        return $PROCESS;
    }
}


1;


lib/Process/Pipeline.pm
package Process::Pipeline;
use 5.008001;
use strict;
use warnings;

use File::Temp ();

our $VERSION = '0.03';

{
    package Process::Pipeline::Process;
    my %SUPPORT_SET = map { $_ => 1 } qw(<  >  >>  2>  2>>  2>&1);
    sub new {
        my $class = shift;
        bless { cmd => [], set => {} }, $class;
    }
    sub cmd {
        my ($self, @arg) = @_;
        if (@arg) {
            if (ref $arg[0] eq 'CODE') {
                $self->{cmd} = $arg[0];
            } else {
                $self->{cmd} = \@arg;
            }
        }
        $self->{cmd};
    }
    sub set {
        my ($self, $key, $value) = @_;
        if ($key) {
            die "Unsupport set '$key'" unless $SUPPORT_SET{$key};
            $self->{set}{$key} = $value;
        }
        $self->{set};
    }
}

{
    package Process::Pipeline::Result::Each;
    sub new { my ($class, %option) = @_; bless {%option}, $class }
    sub status { shift->{status} }
    sub cmd    { shift->{cmd} }
    sub pid    { shift->{pid} }
}

{
    package Process::Pipeline::Result;
    use POSIX ();
    use Process::Status;
    use overload '@{}' => sub { shift->{result} };

    sub new {
        my $class = shift;
        bless {result => [], fh => undef}, $class;
    }
    sub push :method {
        my ($self, $hash) = @_;
        push @$self, $hash;
        $self;
    }
    sub is_success {
        my $self = shift;
        @$self == grep { $_->status->is_success } @$self;
    }
    sub fh {
        my $self = shift;
        $self->{fh} = shift if @_;
        $self->{fh};
    }
    sub wait :method {
        my $self = shift;
        while (grep { !defined $_->status } @$self) {
            my $pid = waitpid -1, POSIX::WNOHANG();
            my $save = $?;
            if ($pid == 0) {
                select undef, undef, undef, 0.01;
            } elsif ($pid == -1) {
                last;
            } else {
                my ($found) = grep { $_->pid == $pid } @$self;
                if (!$found) {
                    warn "waitpid returns $pid, but is not our child!";
                    last;
                }
                $found->{status} = Process::Status->new($save);
            }
        }
        if (my $filename = delete $self->{_filename}) {
            local $!;
            unlink $filename;
        }
        $self;
    }
}

sub new {
    bless { process => [] }, shift;
}

sub push :method {
    my ($self, $callback) = @_;
    my $p = Process::Pipeline::Process->new;
    $callback->($p);
    $self->_push($p);
}

sub _push {
    my ($self, $p) = @_;
    push @{$self->{process}}, $p;
    $self;
}

sub start {
    my ($self, %option) = @_;
    my $n = $#{$self->{process}};
    my @pipe = map { pipe my $read, my $write; [$read, $write] } 0..($n - 1);
    my $close = sub {
        my $i = shift;
        my @close = map { @{$pipe[$_]} } grep { $_ != $i - 1 && $_ != $i } 0..$#pipe;
        $_->close for @close;
    };

    my ($main_out_fh, $main_out_filename);
    my $result = Process::Pipeline::Result->new;
    for my $i (0..$n) {
        my $process = $self->{process}[$i];
        if ($i == $n && !$process->set->{">"} && !$process->set->{">>"}) {
            ($main_out_fh, $main_out_filename) = File::Temp::tempfile(UNLINK => 0);
        }
        my $pid = fork;
        die "fork: $!" unless defined $pid;
        if ($pid == 0) {
            if ($main_out_filename) {
                close $main_out_fh;
                open STDOUT, ">>", $main_out_filename or die $!;
            }
            $close->($i);
            my $read  = $i - 1 >= 0 ? $pipe[$i - 1] : undef;
            my $write = $pipe[$i];
            if ($read) {
                $read->[1]->close;
                open STDIN, "<&", $read->[0];
                $read->[0]->close;
            }
            if ($write) {
                $write->[0]->close;
                open STDOUT, ">&", $write->[1];
                $write->[1]->close;
            }

            my %set = %{$process->set};
            if (my $in = $set{"<"}) {
                open STDIN, "<", $in or die "open $in: $!";
            }
            if (my $out = $set{">"} or my $append = $set{">>"}) {
                my $mode = defined $out ? ">"  : ">>";
                my $file = defined $out ? $out : $append;
                open STDOUT, $mode, $file or die "open $file: $!";
            }
            if (my $out = $set{"2>"} or my $append = $set{"2>>"}) {
                my $mode = defined $out ? ">"  : ">>";
                my $file = defined $out ? $out : $append;
                open STDERR, $mode, $file or die "open $file: $!";
            }
            if (exists $set{"2>&1"}) {
                open STDERR, ">&", \*STDOUT;
            }
            STDOUT->autoflush(1);

            my $cmd = $process->cmd;
            if (ref $cmd eq "CODE") {
                $cmd->();
                exit;
            } else {
                my @cmd = @$cmd;
                exec {$cmd[0]} @cmd;
                exit 255;
            }
        }
        $result->push(Process::Pipeline::Result::Each->new(
            pid => $pid,
            cmd => $process->cmd,
            status  => undef,
        ));
    }
    $_->close for map { @$_ } @pipe;
    if ($main_out_filename) {
        $result->{_filename} = $main_out_filename;
        $result->fh($main_out_fh);
    }
    $result->wait unless $option{async};
    $result;
}

1;
__END__

=encoding utf-8

=head1 NAME

Process::Pipeline - execute processes as pipeline

=head1 SYNOPSIS

In shell:

   $ zcat access.log.gz | grep -v 127.0.0.1 | grep -c POST

In perl5:

  use Process::Pipeline;

  my $pipeline = Process::Pipeline->new
    ->push(sub { my $p = shift; $p->cmd("zcat", "access.log.gz")   })
    ->push(sub { my $p = shift; $p->cmd("grep", "-v", "127.0.0.1") })
    ->push(sub { my $p = shift; $p->cmd("grep", "-c", "POST"       });

  my $result = $pipeline->start;
  if ($result->is_success) {
     my $fh = $result->fh; # output filehandle of $pipeline
     say <$fh>;
  }

In perl5 with DSL:

  use Process::Pipeline::DSL;

  my $pipeline = proc { "zcat", "access.log.gz"   }
                 proc { "grep", "-v", "127.0.0.1" }
                 proc { "grep", "-c", "POST"      };

  my $result = $pipeline->start;
  if ($result->is_success) {
     my $fh = $result->fh; # output filehandle of $pipeline
     say <$fh>;
  }

=head1 DESCRIPTION

Process::Pipeline helps you write a pipeline of processes.

=head1 MOTIVATION

It is known that we should avoid shell-invocation in perl.
But, because the notation of shell is very convenient,
I sometimes find myself invoking shell. Oops.

The main reason for invoking shell in perl is
that perl does not have as convenient notation as shell has.

Process::Pipeline try to give an easy pipeline notation to perl. Why don't you change

    chomp(my $num = `zcat access.log.gz | grep -v 127.0.0.1 | grep -c POST`);

into

    use Process::Pipeline::DSL;
    my $p = proc { "zcat", "access.log.gz"   }
            proc { "grep", "-v", "127.0.0.1" }
            proc { "grep", "-c", "POST"      };
    my $r = $p->start;
    if ($r->is_success) {
      my $fh = $r->fh;
      chomp(my $num = <$fh>);
    }

=head1 METHODS

=head2 new

  my $pipeline = Process::Pipeline->new;

Constructor.

=head2 push

  $pipeline->push(sub {
    my $p = shift;
    $p->cmd("zcat", "access.log.gz");
  });
  $pipeline->push(sub {
    my $p = shift;
    $p->set("2>", "/dev/null");
    $p->cmd("zcat", "access.log.gz");
  });

Push a Process::Pipeline::Process object to the pipeline.

=head2 start

   my $result = $pipeline->start;

Start the pipeline. It returns a Process::Pipeline::Result object.

   my $result = $pipeline->start;
   my $bool   = $reuslt->is_success; # all commands exit successfully
   my $fh     = $reuslt->fh;         # pipeline's output filehandle

=head1 DSL

There is a DSL for Process::Pipeline. Process::Pipeline::DSL exports
C<proc> and C<set> functions, and you can construct pipelines easily.

  use Process::Pipeline::DSL;
  my $p = proc { "git", "archive", "--format=tar", "--prefix=repo/", "HEAD" }
          proc { set ">" => "repo.tar.gz"; "gzip" };
  my $r = $p->start;

=head1 COPYRIGHT AND LICENSE

Copyright 2016 Shoichi Kaji <skaji@cpan.org>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
